packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,319 @@
+			domSourceEditingElementTextarea.addEventListener( 'input', () => {
+				domSourceEditingElementWrapper.dataset.value = domSourceEditingElementTextarea.value;
+			} );
+
+			editingView.change( writer => {
+				const viewRoot = editingView.document.getRoot( rootName );
+
+				writer.addClass( 'ck-hidden', viewRoot );
+			} );
+
+			this._replacedRoots.set( rootName, domSourceEditingElementWrapper );
+
+			this._elementReplacer.replace( domRootElement, domSourceEditingElementWrapper );
+		}
+
+		this._focusSourceEditing();
+	}
+
+	/**
+	 * Restores all hidden editing roots and sets the source data in them.
+	 *
+	 * @private
+	 */
+	_hideSourceEditing() {
+		const editor = this.editor;
+		const editingView = editor.editing.view;
+
+		const data = {};
+
+		for ( const [ rootName, domSourceEditingElementWrapper ] of this._replacedRoots ) {
+			const oldData = editor.data.get( { rootName } );
+			const newData = domSourceEditingElementWrapper.dataset.value;
+
+			// Do not set the data unless some changes has been made in the meantime.
+			// This prevents empty undo steps after switching to the normal editor.
+			if ( oldData !== newData ) {
+				data[ rootName ] = newData;
+			}
+
+			editingView.change( writer => {
+				const viewRoot = editingView.document.getRoot( rootName );
+
+				writer.removeClass( 'ck-hidden', viewRoot );
+			} );
+		}
+
+		this._elementReplacer.restore();
+
+		this._replacedRoots.clear();
+
+		if ( Object.keys( data ).length ) {
+			editor.data.set( data, { batchType: 'default' } );
+		}
+
+		editor.editing.view.focus();
+	}
+
+	/**
+	 * Focuses the textarea containing document source from the first editing root.
+	 *
+	 * @private
+	 */
+	_focusSourceEditing() {
+		const [ domSourceEditingElementWrapper ] = this._replacedRoots.values();
+
+		domSourceEditingElementWrapper.querySelector( 'textarea' ).focus();
+	}
+
+	/**
+	 * Disables all commands.
+	 *
+	 * @private
+	 */
+	_disableCommands() {
+		const editor = this.editor;
+
+		for ( const command of editor.commands.commands() ) {
+			command.forceDisabled( COMMAND_FORCE_DISABLE_ID );
+		}
+	}
+
+	/**
+	 * Clears forced disable for all commands, that was previously set through {@link #_disableCommands}.
+	 *
+	 * @private
+	 */
+	_enableCommands() {
+		const editor = this.editor;
+
+		for ( const command of editor.commands.commands() ) {
+			command.clearForceDisabled( COMMAND_FORCE_DISABLE_ID );
+		}
+	}
+
+	/**
+	 * Adds or removes the `readonly` attribute from textarea from all roots, if document source mode is active.
+	 *
+	 * @param {Boolean} isReadOnly Indicates whether all textarea elements should be read-only.
+	 */
+	_handleReadOnlyMode( isReadOnly ) {
+		if ( !this.isSourceEditingMode ) {
+			return;
+		}
+
+		for ( const [ , domSourceEditingElementWrapper ] of this._replacedRoots ) {
+			domSourceEditingElementWrapper.querySelector( 'textarea' ).readOnly = isReadOnly;
+		}
+	}
+
+	/**
+	 * Checks, if the editor's editable belongs to the editor's DOM tree.
```

Comment:
I'd move it to the private comment inside the method. I'd tell more here what means `true` and `false` from the `_isAllowedToHandleSourceEditingMode`() point of view.

Reply:
Done.
Please check, if it is better now 🙂